src/mlpack/methods/cf/neighbor_search_policies/lmetric_search.hpp

```diff
@@ -0,0 +1,70 @@
+/**
+ * @file lmetric_search.hpp
+ * @author Wenhao Huang
+ *
+ * Search with L_p distance.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_CF_LMETRIC_SEARCH_HPP
+#define MLPACK_METHODS_CF_LMETRIC_SEARCH_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/methods/neighbor_search/neighbor_search.hpp>
+#include <mlpack/core/metrics/lmetric.hpp>
+
+namespace mlpack {
+namespace cf {
+
+/**
+ * Search with L_p distance.
+ *
+ * @tparam TPower Power of metric.
+ */
+template<int TPower>
+class LMetricSearch
+{
+ public:
+  using NeighborSearchType = neighbor::NeighborSearch<
+      neighbor::NearestNeighborSort,
+      metric::LMetric<TPower, true>>;
+
+  /**
+   * @param Set of reference points.
+   */
+  LMetricSearch(const arma::mat& referenceSet) : neighborSearch(referenceSet)
+  { }
+
+  /**
+   * Given a set of query points, find the nearest k neighbors, and return
+   * similarites. Similarities are less than or equal to one.
+   *
+   * @param query A set of query points.
+   * @param k Number of neighbors to search.
+   * @param neighbors Nearest neighbors.
+   * @param similarites Similarities between query point and its neighbors.
+   */
+  void Search(const arma::mat& query, const size_t k,
+      arma::Mat<size_t>& neighbors, arma::mat& similarities)
+  {
+    neighbor.Search(query, k, neighbors, similarities);
+    
+    // Calculate similarities from L_p distance. We restrict that similarities
+    // are not larger than one.
+    similarites = 1.0 / (1.0 + similarities);
+  }
+
+ private:
+  //! NeighborSearch object.
+  NeighborSearchType neighborSearch;
+};
+
+using EuclideanSearch = LMetricSearch<2>;
```

Comment:
Not sure what the idea behind this is, can you elaborate on this one?

Reply:
@zoq The current CF performs neighbor search with Euclidean Distance. I am working to modify that to policy-based design and provide alternatives like `EuclideanSearch`, `CosineSearch`, `PearsonSearch` (maybe others as well). The `Search()` method in these classes return `neighbors`, and `similarities` which is used in calculating interpolation weights (i.e. `GetWeights()` method in interpolation classes).
As to `LMetricSearch`, I guess it is possible that a user may want to try other L_p distance besides L_2 (Euclidean Distance), though I haven't met such cases.

Reply:
Thanks for the clarification, that makes sense, I was just wondering if a user is going to use the alias and if it's a good idea to put it into the cf namespace.